### ISSUE TITLE
Crash caused by invalid image data

### ISSUE THREAD

**Issue by d5xtgr** [WhisperSystems/Signal-Android#5124]
It is possible to cause the application to crash by attempting to send an invalid file via MMS.  To reproduce this, a file of a non-image file type (like PDF) that has been renamed to end in `.jpg` can be selected in the attachment dialogue, but the application crashes when the user attempts to send this message.  Worse, because the crash occurs without successfully sending the bad file, when the application is restarted, it tries to resume the process and crashes again immediately upon startup.  Recovery can be achieved by manually removing the last entry from the tables `mms` and `parts` in `messages.db`.

Signal should gracefully handle being passed illegal data and avoid crashing if it is impossible to send, display, or thumbnail the user's file.

Related issues:
Selecting an SVG file is permitted, but causes crash + crash-on-restart in #5046.
Receiving a vCard (which is not a file type that can be selected for sending) causes crash + crash-on-restart in #3256.
Receiving a corrupted attachment causes crash + crash-on-restart in #2518.

Signal 3.9.1
CyanogenMod 12.1
Moto G falcon

**Comment by neurotensin** [comment]
I created a pull-request to possibly fix to this, as I currently cannot use the app on my phone. (#5169)

**Comment by 2-4601** [comment]
Put your phone in airplane mode, start Signal, delete the message with the issue.

**Comment by neurotensin** [comment]
Thanks! I had this problem before with TextSecure, but forgot. Hence, I decided to patch it...I would rather the exception was handled, but having a default is more secure to limited downstream exploits.

**Comment by moxie0** [comment]
You didn't include a debug log in your issue, but I think this should already be fixed in 3.10

**Comment by neurotensin** [comment]
I did include a log from the handset, but I did not get any "error id" to give you. I resubmitted a bug from the same email address as my handset which was request 4228, so you can probably correlate them and get the log. I patched against today's master, so I'll wait for the next release.

**Comment by moxie0** [comment]
error id? you have to capture the debug log and paste the url here

**Comment by neurotensin** [comment]
how is that done? On the handset there were no options to capture the debug log, I could only  view it.

**Comment by 2-4601** [comment]
http://support.whispersystems.org/hc/en-us/articles/212535838-How-do-I-send-a-debug-log-

**Comment by neurotensin** [comment]
Hi, ok got it. It would be nice to have another way to collect this, as it is not possible if the app crashes...
https://gist.github.com/anonymous/ea45adee0e0a1d286717

**Comment by moxie0** [comment]
looks like a duplicate of #5046

**Comment by neurotensin** [comment]
yes, but my patch uses Glade's fail check, so gives us a chance to intercept, rather than fail. As I mentioned, I think having a remote-triggered error, is a potential attack vector, since we cannot know what other parts of a device might be compromised. In general, I don't trust mime-types, so when Glide fails, a failsafe is to return known Bitmap. I just grabbed an existing icon, but you get the idea.

**Comment by d5xtgr** [comment]
I've confirmed this is fixed in 3.10; the new behaviour is that the fake picture shows up blank and refuses to send.  Thanks to Moxie and his team.

@neurotensin, as has been noted by others, Airplane Mode will allow you to open the app without the crash, but for completeness, I'll describe how I fixed the problem.  Maybe there are circumstances where it's necessary for someone else.

Using adb with root access to the phone, I pull `/data/data/org.thoughtcrime.securesms/databases/messages.db` onto my computer and open the file in sqlitebrowser.  I click on the second tab, which is labelled 'Browse Data', and I select 'mms' from the 'Table' dropdown.  I scroll to the bottom of the page and select the last row.  I click the button labelled 'Delete Record'.  Before saving and closing, I repeat the process for 'part' in the 'Table' dropdown.  Finally, I use adb to push the modified database into the same directory where I found it.
